lib/std/tempfiles.nim

```diff
@@ -103,46 +116,60 @@ proc genTempPath*(prefix, suffix: string, dir = ""): string =
   let dir = getTempDirImpl(dir)
   result = dir / (prefix & randomPathName(nimTempPathLength) & suffix)
 
-proc createTempFile*(prefix, suffix: string, dir = ""): tuple[fd: File, path: string] =
-  ## Creates a new temporary file in the directory `dir`.
+proc createTempFile*(prefix, suffix: string, dir = ""): tuple[cfile: File, path: string] =
+  ## Creates a new temporary file in the directory `dir`, which must exist
+  ## (empty `dir` will resolve to `getTempDir()`).
   ## 
   ## This generates a path name using `genTempPath(prefix, suffix, dir)` and
   ## returns a file handle to an open file and the path of that file, possibly after
   ## retrying to ensure it doesn't already exist.
   ## 
-  ## If failing to create a temporary file, `IOError` will be raised.
+  ## If failing to create a temporary file, `OSError` will be raised.
   ##
   ## .. note:: It is the caller's responsibility to remove the file when no longer needed.
+  runnableExamples:
+    import std/[os, nre]
+    doAssertRaises(OSError): discard createTempFile("", "", "nonexistent")
+    let (cfile, path) = createTempFile("D20210501T170028", "end")
+    cfile.write "foo"
+    cfile.setFilePos 0
+    assert path.lastPathPart.contains(re"^D20210501T170028(\w+)end$")
+    assert readAll(cfile) == "foo"
+    close cfile
+    assert readFile(path) == "foo"
+    removeFile(path)
+  # xxx why does above work without `cfile.flushFile` ?
   let dir = getTempDirImpl(dir)
-  createDir(dir)
   for i in 0 ..< maxRetry:
     result.path = genTempPath(prefix, suffix, dir)
-    try:
-      result.fd = safeOpen(result.path)
-    except OSError:
-      continue
-    return
+    result.cfile = safeOpen(result.path)
+    if result.cfile != nil:
+      return
 
-  raise newException(IOError, "Failed to create a temporary file under directory " & dir)
+  raise newException(OSError, "Failed to create a temporary file under directory " & dir)
 
 proc createTempDir*(prefix, suffix: string, dir = ""): string =
-  ## Creates a new temporary directory in the directory `dir`.
+  ## Creates a new temporary directory in the directory `dir`, which must exist
+  ## (empty `dir` will resolve to `getTempDir()`).
   ##
   ## This generates a dir name using `genTempPath(prefix, suffix, dir)`, creates
   ## the directory and returns it, possibly after retrying to ensure it doesn't
   ## already exist.
   ##
-  ## If failing to create a temporary directory, `IOError` will be raised.
+  ## If failing to create a temporary directory, `OSError` will be raised.
   ##
   ## .. note:: It is the caller's responsibility to remove the directory when no longer needed.
+  runnableExamples:
+    import std/[os, nre]
+    doAssertRaises(OSError): discard createTempDir("", "", "nonexistent")
+    let dir = createTempDir("D20210501T171932", "end")
+    assert dir.lastPathPart.contains(re"^D20210501T171932(\w+)end$")
```

Comment:
```suggestion
    import std/os
    doAssertRaises(OSError): discard createTempDir("", "", "nonexistent")
    let dir = createTempDir("D20210501T171932", "end")
    assert dir.lastPathPart[^3..^1] == "end"
    assert dir.lastPathPart[0..^12] == "D20210501T171932"
```

Reply:
ditto, I've now removed nre import and instead show what the file may look like in comments, and added a test in the ttempfiles